util/linuxfw/nftables_runner.go

```diff
@@ -0,0 +1,1058 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux
+
+package linuxfw
+
+import (
+	"encoding/hex"
+	"fmt"
+	"net"
+	"net/netip"
+	"reflect"
+	"strings"
+
+	"github.com/google/nftables"
+	"github.com/google/nftables/expr"
+	"tailscale.com/net/tsaddr"
+	"tailscale.com/types/logger"
+)
+
+const (
+	TSForwardChain     = "forward"
+	TSInputChain       = "input"
+	TSPostroutingChain = "postrouting"
+)
+
+type nftable struct {
+	Proto  nftables.TableFamily
+	Filter *nftables.Table
+	Nat    *nftables.Table
+}
+
+type nftablesRunner struct {
+	conn *nftables.Conn
+	nft4 *nftable
+	nft6 *nftable
+
+	v6Available    bool
+	v6NATAvailable bool
+}
+
+// decodeHexString decodes a hex string to a byte slice of the given length.
+func decodeHexString(toLen int, hexStr string) ([]byte, error) {
+	if toLen == 0 {
+		return nil, fmt.Errorf("cannot decode hex string to zero length")
+	}
+	if strings.HasPrefix(hexStr, "0x") {
+		hexStr = strings.TrimPrefix(hexStr, "0x")
+	}
+
+	temp, err := hex.DecodeString(hexStr)
```

Comment:
Instead of going via a temporary string using `DecodeString`, can we do this with `hex.Decode` instead? that should save an allocation.

https://pkg.go.dev/encoding/hex#Decode

```go
ret := make([]byte, toLen)
err := hex.Decode(ret, []byte(hexStr))
```

or something

Reply:
Tested on openWRT and Ubuntu

Reply:
Changed. This method only works for hexStr with correct length (filled with leading zeros). Checked new rules, works correctly.